### ISSUE TITLE
Want to use mixpanel on ios app & extension as well as watchos

### ISSUE THREAD

**Issue by yaalaa** [mixpanel/mixpanel-iphone#449]
v2.9.3 podspec seems broken a bit - it lacks for AppExtension subspec.
v2.9.4 isn't accessible (actually pod gives me just v2.9.1, not later).
I can't wait to get extension spec repaired + watchos support.

**Comment by samgreen** [comment]
The pod is now pushed.

**Comment by yaalaa** [comment]
Nice to see v2.9.4 and watchOS support.
It's pity watchOS lib looks like thin client working through WCSession.
Nevertheless I've just tried to integrate it and have faced with an issue:
```
#if defined(MIXPANEL_WATCH_EXTENSION)
@interface Mixpanel (WatchExtensions) <WCSessionDelegate>
- (void)session:(WCSession *)session didReceiveMessage:(NSDictionary<NSString *, id> *)message;
- (void)session:(WCSession *)session didReceiveMessage:(NSDictionary<NSString *, id> *)message replyHandler:(void(^)(NSDictionary<NSString *, id> *replyMessage))replyHandler;
@end
#endif
```
The above is from Mixpanel.h.

````
#if defined(MIXPANEL_WATCH_EXTENSION)
#pragma mark - WatchExtensions
@implementation Mixpanel (WatchExtensions)

- (void)session:(WCSession *)session didReceiveMessage:(NSDictionary<NSString *, id> *)message {

...

}

...

- (void)session:(WCSession *)session didReceiveMessage:(NSDictionary<NSString *, id> *)message replyHandler:(void(^)(NSDictionary<NSString *, id> *replyMessage))replyHandler {

...

}

...

#endif
```
The above is from Mixpanel.m

mixpanel-app doesn't handle WC messages. Am I correct?

I could define `MIXPANEL_WATCH_EXTENSION` explicitly for mixpanel-app lib (_thanks to building from sources_) but it obviously should break Pods project creation.

As far as I understand removing those #if-def braces from both Mixpanel.h & Mixpanel.m should be quite enough to fix the issue.

**Comment by samgreen** [comment]
So we haven't had a chance to better document this process, and I'm really sorry that you're having trouble. You need to define `MIXPANEL_WATCH_EXTENSION` in the pod as well if you want to add support for the watch methods. You can do this by adding the following to your `Podfile`

    post_install do |installer|
	    installer.pods_project.targets.each do |target|
	        target.build_configurations.each do |config|
	            # Disable Mixpanel AB
	            if target.name == 'Mixpanel'
	                config.build_settings['GCC_PREPROCESSOR_DEFINITIONS'] ||= ['$(inherited)']
	                config.build_settings['GCC_PREPROCESSOR_DEFINITIONS'] << 'MIXPANEL_WATCH_EXTENSION=1'
	            end
	        end
	    end
	end

Obviously you will also need to link your application to `WatchConnectivity.framework` to ensure the linker can resolve all symbols.

**Comment by yaalaa** [comment]
I'm now using ObjC.
I'm going to try the way you suggested to get `MIXPANEL_WATCH_EXTENSION` defined but I have to make it smarter a bit - I've got targets which use Mixpanel but have no need in watch connectivity. 

I would suggest you to make another subspec, kinda app_with_watch_support or AppWithWatch or whatever you'd like. That new spec would clone of the main one + define & link to WC.framework. :)

**Comment by samgreen** [comment]
@yaalaa Any luck with that approach?

**Comment by yaalaa** [comment]
I've managed to compile with the following podspec addition:
```
post_install do |installer|
  installer.pods_project.targets.each do |target|
    if target.name == "Pods-<mytarget>-Mixpanel"
      target.build_configurations.each do |config|
        # Enable Mixpanel watch support
        config.build_settings['GCC_PREPROCESSOR_DEFINITIONS'] ||= ['$(inherited)']
        config.build_settings['GCC_PREPROCESSOR_DEFINITIONS'] << 'MIXPANEL_WATCH_EXTENSION=1'
      end
    end
  end
end
```

Also in sources I do the following:
```
#if defined(NEED_WATCH)
#define MIXPANEL_WATCH_EXTENSION
#endif
#import <Mixpanel/Mixpanel.h>
```

Unfortunately, there's another issue.
Take a look at Mixpanel.h:
```
- (void)session:(WCSession *)session didReceiveMessage:(NSDictionary<NSString *, id> *)message replyHandler:(void(^)(NSDictionary<NSString *, id> *replyMessage))replyHandler {
    NSString *messageType = [Mixpanel messageTypeForWatchSessionMessage:message];
    if (messageType) {
        if ([messageType isEqualToString:@"track"]) {
            [[Mixpanel sharedInstance] track:message[@"event"] properties:message[@"properties"]];
        }
        replyHandler(@{ @"success": @YES });
    } else {
        replyHandler(@{ @"success": @NO, @"message": @"Message is not a mixpanel message" });
    }
}
```

The code above refers to `Mixpanel.sharedInstance` which is not the case for me. I have two *Mixpanel apps* and maintain instances by myself.

BTW, `Mixpanel session:didReceiveMessage:replyHandler` seems to be instance method. Is it necessary to call `Mixpanel.sharedInstance`?

**Comment by samgreen** [comment]
@yaalaa Thanks for your detailed feedback. Happy to hear it's (almost) working for you. The use of `sharedInstance` is an oversight from the original watchOS implementation. We'll be sure to get that removed in the next release.

Just so you're aware, adding:

    #if defined(NEED_WATCH)
    #define MIXPANEL_WATCH_EXTENSION
    #endif
    #import <Mixpanel/Mixpanel.h>

is not necessary. If you define this project wide, either with a precompiled header, or by adding a `GCC_PREPROCESSOR_DEFINITIONS` in your project it won't break anything. Adding this define just _adds_ support for the watchOS communication methods, it doesn't remove *any* functionality from the SDK.

I'll update this ticket when the PR is opened, and again when the new release is out.

**Comment by samgreen** [comment]
The fix required from this issue will be in v2.9.5

**Comment by samgreen** [comment]
This is resolved in v2.9.5 which is now released.
